Definition of the Camera AV StreamManagement Cluster.

pidarped opened the pull request:
XML, json and ZAP generated files.

Spec in `cameras` branch.
Draft PR: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10004

pidarped committed:
Definition of the Camera AV StreamManagement Cluster.

Define StreamTypeEnum in global-enums.xml, moving from WebRTC Provider
cluster and AV Stream Management.

Spec in `cameras` branch.
Draft PR: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10004

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+
+  <struct name="VideoStreamStruct" apiMaturity="provisional">
+    <cluster code="0x0551"/>
+    <item fieldId="0" name="VideoStreamID" type="int16u"/>
```
Spec comment: the spec has lots of "MS" in the Default comment, which probably don't make sense.  But since in practice the XML ignores those anyway, it's not a problem for this PR.

pidarped replied:
MS values for these attributes have been removed in the Spec.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+    <globalAttribute code="0xFFFD" side="either" value="1"/>
+    <attribute code="0x0000" side="server" define="MAX_CONCURRENT_VIDEO_ENCODERS" type="int8u" reportable="true" optional="true">
+      <description>MaxConcurrentVideoEncoders</description>
+      <access op="read" privilege="administer"/>
```
Some spec questions here about whether these should be A or M, but the XML matches the spec, ok.

pidarped replied:
Updated the Access column for all fields in the spec and incorporated in the xml.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+      <access op="read" privilege="administer"/>
+    </attribute>
+    <attribute code="0x0020" side="server" define="SPKR_ON_OFF" type="boolean" writable="true" optional="true">
+      <description>SpkrOnOff</description>
```
Spec comment: here and for the other speaker attributes, should just spell out "Speaker" instead of the weird abbreviation.

pidarped replied:
Expanded to Speaker

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+      <access op="write" privilege="administer"/>
+    </attribute>
+    <attribute code="0x0024" side="server" define="MIC_ON_OFF" type="boolean" writable="true" optional="true">
+      <description>MicOnOff</description>
```
And should consider whether "Microphone" should be shortened.

Also "OnOff" is bad for a boolean, since it's not immediately clear whether true is on or off.

pidarped replied:
Kept `Mic` but removed `OnOff`.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+      <access op="write" privilege="administer"/>
+    </attribute>
+    <attribute code="0x0028" side="server" define="MIC_AGC" type="boolean" writable="true" optional="true">
+      <description>MicAGC</description>
```
Spec comment: Consider calling this MicAGCEnabled for clarity.

pidarped replied:
Done.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+      <access op="write" privilege="administer"/>
+    </attribute>
+    <attribute code="0x002E" side="server" define="STATUS_LIGHT" type="boolean" writable="true" optional="true">
+      <description>StatusLight</description>
```
Spec comment: This really does not seem related to stream management, and seems like it would be better handled by a level control and/or On/Off cluster for the light on a separate endpoint.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+      <arg id="1" name="Resolution" type="VideoResolutionStruct"/>
+    </command>
+
+    <command code="0x06" source="client" name="VideoStreamDeallocate" optional="true">
```
This command allows one client to deallocate a stream a different client, possibly on a different fabric, allocated, right?  Is this intentional?

pidarped replied:
Yes, this cluster is a global cluster across fabrics. And the streams are shared/reused. Admins are allowed to de-allocate provided the ref-count on the stream is 0.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,578 @@
+
+    <command code="0x0F" source="client" name="SetImageFlipHorizontal" optional="true">
+      <description>The data fields for this command SHALL be as follows: </description>
+      <arg id="0" name="OnOff" type="boolean"/>
```
Again, "OnOff" is a bad name for a boolean.

And yes, the On/Off cluster has it.  We should not copy it.

pidarped replied:
Changed to Enabled.

bzbarsky-apple commented on src/app/zap-templates/zcl/zcl-with-test-extensions.json:
```diff
@@ -679,7 +680,60 @@
-        "Service Area": ["CurrentArea", "EstimatedEndTime", "FeatureMap"]
+        "Service Area": ["CurrentArea", "EstimatedEndTime", "FeatureMap"],
+        "Camera AV Stream Management": [
+            "MaxConcurrentVideoEncoders",
```
This and the comment about zap_cluster_list are the only actual non-spec issues here, by the way, and this can land once those are addressed.

gmarcosb commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
You should copy the one from https://github.com/project-chip/connectedhomeip/blob/6d83b821e6d73111f594708af790579ca90ad382/src/app/zap-templates/zcl/data-model/draft/camera-av-stream-management-cluster.xml as it contains some things that are missing here (e.g. the new pointer to the XSD with the new alchemy, description field, etc.)

pidarped replied:
I also had some new changes here related to the definition of the VideoResolution struct. Also, we have a way to define the derived type in chip-types.xml(and a couple of places) but might need to modify the spec too. So, for now, have hardcoded to int16u.

gmarcosb replied:
Ah I missed the spec has an updated name for the struct 👍

Copying the description field & running with the latest alchemy seems like the best approach then 👍 (I've had to do it a couple times as well after applying some of the suggested changes, e.g. `Id` -> `ID`)

pidarped replied:
Regenerated using the latest Alchemy

gmarcosb commented on src/app/zap-templates/zcl/data-model/chip/global-enums.xml:
```diff
@@ -34,6 +34,18 @@ TODO: Make these structures global rather than defining them for each cluster.
     <item name="Automatic" value="0x03"/>
   </enum>
+
+  <enum name="StreamTypeEnum" type="enum8">
```
This fixes https://github.com/project-chip/connectedhomeip/issues/35838

You'll need to update [webrtc-provider-cluster.xml](https://github.com/project-chip/connectedhomeip/blob/ca79d5910a50e6fb2f82748bd1540153d98601b6/src/app/zap-templates/zcl/data-model/chip/webrtc-provider-cluster.xml#L19) & remove from there

pidarped replied:
That is also done with this PR.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,576 @@
+      <description>NightVisionCapable</description>
+      <access op="read" privilege="administer"/>
+    </attribute>
+    <attribute code="0x0004" side="server" define="MIN_VIEW_PORT_WIDTH" type="VideoResolutionStruct" reportable="true" optional="true">
```
Spec issue: using VideoResolutionStruct as the type for MinViewPortWidth and MinViewPortHeight makes no sense to me....

pidarped replied:
Fixed in spec and incorporated in xml. The 2 attributes are collapsed to 1 attribute MinViewport of type VideoResolutionStruct(containing width and height).

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,576 @@
+      <access op="read" privilege="administer"/>
+      <access op="write" privilege="administer"/>
+    </attribute>
+    <attribute code="0x0024" side="server" define="MIC_STATUS" type="boolean" writable="true" optional="true">
```
Spec issue: What does a boolean "MicStatus" mean?  What does true mean?  What does false mean?

This needs a better name.

pidarped replied:
Captured comment in spec.

bzbarsky-apple commented on src/app/zap-templates/zcl/data-model/chip/camera-av-stream-management-cluster.xml:
```diff
@@ -0,0 +1,576 @@
+      <access op="read" privilege="administer"/>
+      <access op="write" privilege="administer"/>
+    </attribute>
+    <attribute code="0x002E" side="server" define="STATUS_LIGHT" type="boolean" writable="true" optional="true">
+      <description>StatusLight</description>
```
OK, if this stays, the same spec comment: What do "true" and "false" mean for an attribute called StatusLight?  Is this about whether it's on?  About whether it's _allowed_ to be on?  Something else entirely?

Needs better naming.

pidarped replied:
Captured comment in spec.

bzbarsky-apple reviewed (approved):
Approving, but please make sure the spec comments, both from this review and the previous one, are captured in the spec.

pidarped committed:
Keep the jclass and its corresponding constructor jmethod inside a scope block.

This is to prevent duplicate declaration error when multiple
variables of the same struct type are present.

pidarped committed:
Auto-generated files from zap-regen for Camera AV Stream Management.
